### DISCUSSION TITLE
`timer_create`, `timer_settime` and `timer_delete` syscall implementation

### DISCUSSION THREAD

**Discussion by ghost** [gramineproject/gramine#1242] [Q&A] [answered]
Started prototyping a bit today to add timer support and ran into two issues.

1. Can we use a global `timerid` counter in `libos_utils.c`? Or would we want a separate handle map inside `libos_handle.c`?
2. The syscall `alarm` seems to be using the `_PalSystemTimeQuery` OCALL implemented in `pal_misc.c` which only supports `CLOCK_REALTIME`. Can we continue ignoring other clock types? Is there an alternative clock source for the enclave?

**Comment by dimakuv** [comment, marked as answer]
> 1. Can we use a global `timerid` counter in `libos_utils.c`? Or would we want a separate handle map inside `libos_handle.c`?

I think you can use a global counter, instead of adding a new handle map. There are two reasons for this:
1. Simplicity. It looks like timer IDs are implemented as [rather complicated numbers in Linux](https://elixir.bootlin.com/linux/latest/source/include/linux/posix-timers.h#L13), but without any connections to other handles (file descriptions). So we can do something similar -- just have a global counter, for the sake of simplicity and self-containment.
2. POSIX timers are process-internal, i.e., they do not propagate over `fork()` and `execve()`. This is described in the [man page](https://man7.org/linux/man-pages/man2/timer_create.2.html):

   > Timers are not inherited by the child of a [fork(2)](https://man7.org/linux/man-pages/man2/fork.2.html), and are disarmed and deleted during an [execve(2)](https://man7.org/linux/man-pages/man2/execve.2.html).

But I would prefer to put *all* POSIX-timer functionality, including this global `timerid` counter, in either of these:
- https://github.com/gramineproject/gramine/blob/master/libos/src/sys/libos_time.c (preferred by me)
- https://github.com/gramineproject/gramine/blob/master/libos/src/sys/libos_posix_time.c (new file, if you feel like it deserves its own file)

But definitely I don't like the idea of using `libos_utils.c` for POSIX timers (or parts of POSIX timers).

---

> 2. The syscall `alarm` seems to be using the `_PalSystemTimeQuery` OCALL implemented in `pal_misc.c` which only supports `CLOCK_REALTIME`. Can we continue ignoring other clock types? Is there an alternative clock source for the enclave?

Your understanding is correct. There is **no alternative** clock source for the enclave.

I suggest to follow the logic as here: https://github.com/gramineproject/gramine/blob/0119fd23ebbc892b33868e9ea9296ba6e8061685/libos/src/sys/libos_sleep.c#L76

In other words, emulate all clocks as `CLOCK_REALTIME` and output a warning about this.

---

> Based on my current understanding adding timer support will require `timer_create`, `timer_settime` and `timer_delete` to be implemented.

Yes. Actually, according to the [man page](https://man7.org/linux/man-pages/man2/timer_create.2.html), you'll need to implement the following syscalls:

*  timer_create(): Create a timer.

*  [timer_settime(2)](https://man7.org/linux/man-pages/man2/timer_settime.2.html): Arm (start) or disarm (stop) a timer.

*  [timer_gettime(2)](https://man7.org/linux/man-pages/man2/timer_gettime.2.html): Fetch the time remaining until the next  expiration of a timer, along with the interval setting of the timer.

*  [timer_getoverrun(2)](https://man7.org/linux/man-pages/man2/timer_getoverrun.2.html): Return the overrun count for the last timer expiration.

*  [timer_delete(2)](https://man7.org/linux/man-pages/man2/timer_delete.2.html): Disarm and delete a timer.

By the way, I would recommend to start with only `SIGEV_NONE` and `SIGEV_SIGNAL` support in your first PR. The other types (`SIGEV_THREAD` and `SIGEV_THREAD_ID`) seem more complicated and may be added in a future PR (if required).

I think I already mentioned this in the past, but please use the following logic as your guide in POSIX-timers implementation:
- https://github.com/gramineproject/gramine/blob/master/libos/src/sys/libos_time.c
- https://github.com/gramineproject/gramine/blob/master/libos/src/sys/libos_sleep.c

**Reply by ghost** [reply]
Thanks for the in-depth response!
